### ISSUE TITLE
Java NVIDIA GPU Inference Support

### ISSUE THREAD

**Issue by jxtps** [pytorch/pytorch#63581]
## 🚀 Feature
The [Java Demo](https://github.com/pytorch/java-demo) now supports Linux, Mac, and Windows. That's great. However, looking at [Device.java](https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/src/main/java/org/pytorch/Device.java) in the underlying [pytorch_android library](https://github.com/pytorch/pytorch/tree/master/android/pytorch_android), it appears to only support CPU and VULKAN (= mobile android GPU). 

Please add support for Java NVIDIA GPU inference. 

## Motivation
Our entire web stack is Java/JVM based. It would be massively helpful to be able to deploy pytorch models for Java-based cloud NVIDIA GPU inference. 

## Pitch

Pretty please? Java is a pretty widely used language, and better support for inference in it would broaden the appeal of PyTorch. 

## Alternatives

[Export the pytorch model to an ONNX model](https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html) and then use [JavaCPP ONNX presets](https://github.com/bytedeco/javacpp-presets/tree/master/onnxruntime). _This adds undesirable steps & complexity to the pipeline_

Unfortunately the JavaCPP PyTorch presets don't (yet?) support module loading: see https://github.com/bytedeco/javacpp-presets/issues/1068

**Comment by dreiss** [comment]
This can be done by putting CPU->CUDA and CUDA->CPU copies directly in your model before exporting.  See https://github.com/dreiss/java-demo/issues/2

**Comment by jxtps** [comment]
Interesting! 

Sounds like I'd have to export two separate models then if I want both CPU and GPU support?

**Comment by dreiss** [comment]
Yes.  Or if you use torch.jit.script, you can conditionally copy to CUDA depending on an extra argument passed to the model.

**Comment by jxtps** [comment]
Wow. That's nuts & awesome. 

Would it be possible to do some `if torch.cuda_is_available():` type magic, or would that be evaluated & flattened at model export time, rather than re-run at runtime inside torchscript?

**Comment by dreiss** [comment]
I believe.  It would be flattened if you used `torch.jit.trace`, but not if you use `torch.jit.script`.  If you need to use tracing, you could try tracing your model first, then putting the traced model inside a new wrapper module with the `cuda_is_available` logic, and running `torch.jit.script` on that.

**Comment by dreiss** [comment]
Alternatively, you could write a path to pytorch_android that supports mapping to a CUDA device.  I think we could merge that as long as it's only enabled for CUDA builds.

**Comment by jxtps** [comment]
> write a path to pytorch_android that supports mapping to a CUDA device

What does that mean? I have a very superficial knowledge of pytorch, so I'm not familiar with the vernacular here.

**Comment by jxtps** [comment]
When I try to use `torch.cuda.is_available()` with `torch.jit.script` I get the following error: 
```
```
